Mathlib/Algebra/Polynomial/Laurent.lean

```diff
@@ -512,6 +510,78 @@ theorem isLocalization : IsLocalization (Submonoid.powers (X : R[X])) R[T;T⁻¹
       rintro rfl
       exact ⟨1, rfl⟩ }
 
+theorem mk'_mul_T (p : R[X]) (n : ℕ) :
+  IsLocalization.mk' R[T;T⁻¹] p (⟨X^n,n,rfl⟩ : Submonoid.powers (X : R[X])) * T n =
+    toLaurent p := by
+  rw [←toLaurent_X_pow, ←algebraMap_eq_toLaurent, IsLocalization.mk'_spec, algebraMap_eq_toLaurent]
+
+theorem mk'_eq (p : R[X]) (n : ℕ) : IsLocalization.mk' R[T;T⁻¹] p
+  (⟨X^n,n,rfl⟩ : Submonoid.powers (X : R[X])) = toLaurent p * T (-n) := by
+  rw [←IsUnit.mul_left_inj (isUnit_T n), mul_T_assoc, neg_add_cancel, T_zero, mul_one]
+  exact mk'_mul_T p n
+
+theorem mk'_one_X_pow (n : ℕ) : IsLocalization.mk' R[T;T⁻¹] 1
+  (⟨X^n,n,rfl⟩ : Submonoid.powers (X : R[X])) = T (-n) := by
+  rw [mk'_eq 1 n, toLaurent_one, one_mul]
+
+theorem mk'_one_X : IsLocalization.mk' R[T;T⁻¹] 1
+  (⟨X,1,pow_one X⟩ : Submonoid.powers (X : R[X])) = T (-1) := by
+  convert mk'_one_X_pow 1
+  exact (pow_one X).symm
+
+/-- Given a ring homomorphism `f : R →+* S` and a unit `x` in `S`, the induced homomorphism
+`R[T;T⁻¹] →+* S` sending `T` to `x` and `T⁻¹` to `x⁻¹`. -/
+def eval₂ : R[T;T⁻¹] →+* S :=
```

Comment:
I think there are three options here:

1. Take an argument `x : Units S`
2. Take arguments `x : S` and `hx : IsUnit x` (the current choice)
3. Take an argument `x : S` and use junk values when `x` is not a unit.

Option (1) gives the user good control over the inverse of `x`, and so we can get nice computation lemmas for the image of `T ^ (-n)`. But the price is that the user has to bundle things in to a term of type `Units`.

Option (3) means that the user doesn't have to worry about proof obligations while evaluating at `x`. But occasional lemmas might need `IsUnit x` as condition.

Option (2) seems to sit in the middle, meaning that the user has to always provide the proof obligation, without getting the benefit of good control about the inverse of `x`.

At the moment, I think my preference would be option (1), then (3), then (2).
What do you think?

Reply:
I agree that Option (1) is better.